cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceRuntimeService.java

```diff
@@ -496,28 +498,42 @@ private Location createTempLocationDirectory() throws IOException {
    */
   @SuppressWarnings("unchecked")
   private void beforeSubmit(final Job job) throws TransactionFailureException {
+
+    // AbstractMapReduce implements final initialize(context) and requires subclass to
+    // implement initialize(), whereas programs that directly implement MapReduce have
+    // the option to override initialize(context) (if they implement ProgramLifeCycle)
+    final TransactionControl txControl = mapReduce instanceof AbstractMapReduce
+      ? Transactions.getTransactionControl(TransactionControl.IMPLICIT, AbstractMapReduce.class,
+                                           mapReduce, "initialize")
+      : mapReduce instanceof ProgramLifecycle
+      ? Transactions.getTransactionControl(TransactionControl.IMPLICIT, MapReduce.class,
+                                           mapReduce, "initialize", MapReduceContext.class)
+      : TransactionControl.IMPLICIT;
+
+    if (TransactionControl.EXPLICIT == txControl) {
+      ClassLoader oldClassLoader = ClassLoaders.setContextClassLoader(job.getConfiguration().getClassLoader());
+      try {
+        doInitialize();
+      } catch (Exception e) {
+        throw Throwables.propagate(e);
+      } finally {
+        ClassLoaders.setContextClassLoader(oldClassLoader);
+      }
+    }
     TransactionContext txContext = context.getTransactionContext();
     Transactions.execute(txContext, "initialize", new Callable<Void>() {
       @Override
       public Void call() throws Exception {
         ClassLoader oldClassLoader = ClassLoaders.setContextClassLoader(job.getConfiguration().getClassLoader());
         try {
-          context.setState(new ProgramState(ProgramStatus.INITIALIZING, null));
-          if (mapReduce instanceof ProgramLifecycle) {
-            ((ProgramLifecycle) mapReduce).initialize(context);
-          } else {
-            mapReduce.beforeSubmit(context);
+          if (TransactionControl.IMPLICIT == txControl) {
+            doInitialize();
           }
-          // once the initialize method is executed, set the status of the MapReduce to RUNNING
-          context.setState(new ProgramState(ProgramStatus.RUNNING, null));
-
           // set input/outputs info, and get one of the configured mapper's TypeToken
           TypeToken<?> mapperTypeToken = setInputsIfNeeded(job);
           setOutputsIfNeeded(job);
-
           setOutputClassesIfNeeded(job, mapperTypeToken);
           setMapOutputClassesIfNeeded(job, mapperTypeToken);
```

Comment:
So these couple `setXXX` calls don't need to be inside transaction, or do they? I was a bit confused looking at this method as I expect something more like this:

```java
TransactionControl txControl = // determine tx control like above
if (txControl == EXPLICIT) {
  doInitialize();  // assuming the setting of ClassLoader is inside the doInitialize method
} else {
  Transactions.execute(...., new Callable<Void>() {
    public Void call() throws Exception {
      doInitialize();
    }
  });
}

// Set MR job properties based on result from the initialize call
TypeToken<?> mapperTypeToken = setInputsIfNeeded(job);
......
```